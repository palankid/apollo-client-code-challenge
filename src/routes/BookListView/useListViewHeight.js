import { useEffect, useState } from 'react';

import { debounce } from '../../utils/execution.utils';

const useListViewHeight = (headerRef) => {
    const [listViewHeight, setListViewHeight] = useState(0);

    const onWindowResize = debounce(() => {
        const headerHeight = headerRef?.current?.getBoundingClientRect?.().height || 0;
        const listViewHeight = window.innerHeight - headerHeight;
        setListViewHeight(listViewHeight);
    }, 100);

    useEffect(() => {
        window.addEventListener('resize', onWindowResize);
        onWindowResize();

        () => window.removeEventListener('resize', onWindowResize)
    }, []);

    return listViewHeight;
};

export default useListViewHeight;
