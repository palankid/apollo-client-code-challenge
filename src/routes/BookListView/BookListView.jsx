import React, { useRef, useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';

import { useGetBookList } from '../../operations/queries/getBookList';

import Header from './components/Header';
import ListView from './components/ListView';
import useListViewHeight from './useListViewHeight';

const BookListView = () => {
    const history = useHistory();
    const headerRef = useRef(null);
    const listViewHeight = useListViewHeight(headerRef);
    const { loading, error, data } = useGetBookList();

    if (error) return <p>Error: {error}</p>;

    return (
        <div className="book-list-view">
            <div ref={headerRef}>
              <Header />
            </div>
            <ListView
              listViewHeight={listViewHeight}
              books={data?.books}
              loading={loading}
            />
        </div>
    );
};

export default BookListView;
