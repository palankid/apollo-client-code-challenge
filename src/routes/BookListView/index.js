import BookListView from './BookListView';

export default BookListView;
