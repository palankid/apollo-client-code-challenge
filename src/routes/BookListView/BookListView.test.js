import React from 'react';
import { shallow } from 'enzyme';

import BookListView from './BookListView';

jest.mock('../../operations/queries/getBookList', () => ({
    useGetBookList: () => ({
        loading: false,
        error: null,
        data: {}
    })
}));
jest.mock('./useListViewHeight', () => ({
    __esModule: true,
    default: () => 0,
}));

describe('BookListView', () => {
    it('should test that component is rendered and not changed unintentionally', () => {
        const wrapper = shallow(
            <BookListView />
        );

        expect(wrapper).toMatchSnapshot();
    });
});
