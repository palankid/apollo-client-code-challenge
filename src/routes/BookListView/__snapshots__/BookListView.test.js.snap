// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookListView should test that component is rendered and not changed unintentionally 1`] = `
<div
  className="book-list-view"
>
  <div>
    <Header />
  </div>
  <BookList
    books={Array []}
    listViewHeight={0}
    loading={false}
  />
</div>
`;
